starboard/linux/shared/pre_app_recommendation_service.cc

```diff
@@ -0,0 +1,183 @@
+
+  // Default constructor.
+  PreAppRecommendationsPlatformServiceImpl() = default;
+
+} PreAppRecommendationsPlatformServiceImpl;
+
+// Use HTTP status code in response to YouTube application's method recommend
+// call.
+const char kSuccess[] = "200";
+const char kBadRequest[] = "400";
+
+bool Has(const char* name) {
+  // Check if platform has service name.
+  return strcmp(name, kPreappRecommendationServiceName) == 0;
+}
+
+PlatformServiceImpl* Open(void* context,
+                          ReceiveMessageCallback receive_callback) {
+  SB_DCHECK(context);
+  SB_LOG(INFO) << "Open() service created: "
+               << kPreappRecommendationServiceName;
+
+  return new PreAppRecommendationsPlatformServiceImpl(context,
+                                                      receive_callback);
+}
+
+void Close(PlatformServiceImpl* service) {
+  // Function Close shouldn't manually delete PlatformServiceImpl pointer,
+  // because it is managed by unique_ptr in Platform Service.
+  SB_LOG(INFO) << "Perform actions before gracefully shutting down the service";
+}
+
+std::string extractJsonValue(const std::string& jsonLikeString,
+                             const std::string& key) {
+  std::string result;
+  std::size_t start = jsonLikeString.find("\"" + key + "\"");
+
+  if (start != std::string::npos) {
+    start = jsonLikeString.find(":", start);
+    if (start != std::string::npos) {
+      start = jsonLikeString.find_first_of("\"", start);
+      if (start != std::string::npos) {
+        ++start;  // Skip the opening quote
+        std::size_t end = jsonLikeString.find("\"", start);
+        if (end != std::string::npos) {
+          result = jsonLikeString.substr(start, end - start);
+        }
+      }
+    }
+  }
+
+  return result;
+}
+
+void* Send(PlatformServiceImpl* service,
+           void* data,
+           uint64_t length,
+           uint64_t* output_length,
+           bool* invalid_state) {
+  SB_DCHECK(service);
+  SB_DCHECK(data);
+  SB_DCHECK(output_length);
+
+  char message[length + 1];
+  std::memcpy(message, data, length);
+  message[length] = '\0';
+
+  std::string response = "";
+
+  // TODO: Replace extractJsonValue function with a robust JSON parsing
+  // library for production use. The current implementation has limited
+  // capabilities that nested object isn't supported.
+  std::string method_name = extractJsonValue(message, "method");
+
+  if (method_name.length() == 0) {
+    SB_LOG(ERROR) << "Could not extract method value from the input JSON file:"
+                  << message;
+  }
+
+  // When method_name = getPartnerId, platform returns partner Id to get
+  // authenticated by YouTube app.
+  if (method_name == "getPartnerId") {
+    // Populate Partner Id in the response.
+    // Partner Id will be used by YouTube app to authenticate platforms and
+    // retrieve specified topics for a platform.
+    auto partner_id = "\"dummmy_partner_id\"";
```

Comment:
This looks more like a configuration constant, so it should probably be moved to an easy to find place somewhere at the top of the module. E.g. ````

```
// Partner Id will be used by YouTube app to authenticate platforms and
// retrieve specified topics for a platform. 
// Replace this with your platforms ID in production code.
const char kPartnerId = "dummy_partner_id";
```